File: backend/app/schemas/cms.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ArticleType

# §1 asks for the full desk form. These are shared by the create and patch
# bodies; ArticlePatch differs only in that everything is optional, so the
# PATCH semantics stay "change exactly what was sent" (`exclude_unset`).


class ArticleWrite(BaseModel):
    title_te: str = Field(min_length=3, max_length=400)
    title_en: str | None = Field(default=None, max_length=400)
    sub_title_te: str | None = Field(default=None, max_length=500)
    summary_te: str | None = Field(default=None, max_length=1000)
    body: dict[str, Any] | None = None

    # placement
    category_id: int | None = None
    subcategory_id: int | None = None
    district_id: int | None = None
    mandal_id: int | None = None
    locality_id: int | None = None

    # media (§1: image, gallery, video)
    hero_media_id: int | None = None
    gallery_media_ids: list[int] | None = Field(default=None, max_length=40)
    video_id: int | None = None
    video_youtube_url: str | None = Field(
        default=None, max_length=300,
        description="Pasting a YouTube URL creates/reuses a Video row and links it.")

    # taxonomy
    tags: list[str] | None = Field(
        default=None, max_length=25,
        description="Tag names or slugs. Unknown names are created as topic tags.")

    # byline and provenance
    byline_te: str | None = Field(default=None, max_length=200)
    author_id: int | None = None
    source_type: str = Field(default="own", pattern=r"^(own|agency|contributed|syndicated)$")
    source_credit: str | None = Field(default=None, max_length=200)
    article_type: ArticleType | None = None

    # flags
    is_breaking: bool = False
    is_exclusive: bool = False
    is_featured: bool = False
    voice_enabled: bool = True

    # SEO (§1)
    slug: str | None = Field(default=None, max_length=180, pattern=r"^[a-z0-9][a-z0-9-]*$")
    seo_title: str | None = Field(default=None, max_length=200)
    seo_description: str | None = Field(default=None, max_length=400)
    canonical_url: str | None = Field(default=None, max_length=500)

    # timing (§1 publish date and time)
    scheduled_at: datetime | None = None
    expires_at: datetime | None = None
    breaking_until: datetime | None = None

    # §8 / §9 placement, chosen while writing. Applied when the story goes live.
    pin_home_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)
    pin_trending_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)


class ArticlePatch(BaseModel):
    title_te: str | None = Field(default=None, min_length=3, max_length=400)
    title_en: str | None = Field(default=None, max_length=400)
    sub_title_te: str | None = Field(default=None, max_length=500)
    summary_te: str | None = Field(default=None, max_length=1000)
    body: dict[str, Any] | None = None

    category_id: int | None = None
    subcategory_id: int | None = None
    district_id: int | None = None
    mandal_id: int | None = None
    locality_id: int | None = None

    hero_media_id: int | None = None
    gallery_media_ids: list[int] | None = Field(default=None, max_length=40)
    video_id: int | None = None
    video_youtube_url: str | None = Field(default=None, max_length=300)

    tags: list[str] | None = Field(default=None, max_length=25)

    byline_te: str | None = Field(default=None, max_length=200)
    author_id: int | None = None
    source_type: str | None = Field(default=None, pattern=r"^(own|agency|contributed|syndicated)$")
    source_credit: str | None = Field(default=None, max_length=200)
    article_type: ArticleType | None = None

    is_breaking: bool | None = None
    is_exclusive: bool | None = None
    is_featured: bool | None = None
    voice_enabled: bool | None = None

    slug: str | None = Field(default=None, max_length=180, pattern=r"^[a-z0-9][a-z0-9-]*$")
    seo_title: str | None = Field(default=None, max_length=200)
    seo_description: str | None = Field(default=None, max_length=400)
    canonical_url: str | None = Field(default=None, max_length=500)

    scheduled_at: datetime | None = None
    expires_at: datetime | None = None
    breaking_until: datetime | None = None

    # §8 / §9 placement, chosen while writing. Applied when the story goes live.
    pin_home_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)
    pin_trending_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)


class TransitionIn(BaseModel):
    note: str | None = Field(default=None, max_length=1000)
    #: `publish` accepts a future time, which schedules instead of publishing.
    scheduled_at: datetime | None = None


class CmsMediaRef(BaseModel):
    id: int
    url: str
    alt_te: str | None = None
    credit: str | None = None
    width: int | None = None
    height: int | None = None


class CmsVideoRef(BaseModel):
    id: int
    youtube_id: str
    title_te: str
    thumbnail_url: str


class CmsAudioRef(BaseModel):
    """§19 — whichever rendition the article currently uses, generated or
    uploaded. `provider == 'upload'` is a file an editor attached by hand."""

    id: int
    url: str | None
    mime: str
    duration_sec: int
    provider: str
    status: str


class CmsTagRef(BaseModel):
    id: int
    slug: str
    name_te: str
    name_en: str


class CmsArticleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    short_id: str
    slug: str
    title_te: str
    title_en: str | None
    sub_title_te: str | None = None
    summary_te: str | None
    body: dict[str, Any] | None
    category_id: int | None
    subcategory_id: int | None = None
    district_id: int | None
    mandal_id: int | None
    locality_id: int | None = None
    author_id: int | None
    byline_te: str | None
    source_type: str
    source_credit: str | None
    article_type: str = ArticleType.NORMAL.value
    status: str
    workflow_state: str
    is_breaking: bool
    is_exclusive: bool
    is_featured: bool = False
    voice_enabled: bool = True
    ai_generated: bool
    hero_media_id: int | None = None
    video_id: int | None = None
    audio_asset_id: int | None = None
    seo_title: str | None = None
    seo_description: str | None = None
    canonical_url: str | None = None
    approved_by: int | None
    approved_at: datetime | None
    published_at: datetime | None
    scheduled_at: datetime | None = None
    expires_at: datetime | None = None
    breaking_until: datetime | None = None
    pin_home_minutes: int | None = None
    pin_trending_minutes: int | None = None
    updated_at: datetime

    # Resolved for the editor form; not columns.
    hero_media: CmsMediaRef | None = None
    gallery: list[CmsMediaRef] = []
    video: CmsVideoRef | None = None
    tags: list[CmsTagRef] = []
    audio: CmsAudioRef | None = None
    #: Live pins for this article, so the form shows what is actually running
    #: rather than only what was requested.
    active_pins: list[dict[str, Any]] = []


class PlacementIn(BaseModel):
    """§8 / §9 — set (or clear) the home and Top-trending pins for one article.

    Separate from ArticlePatch because a published article cannot be edited,
    but its placement very much can: the story leading the front page is a
    decision made *after* publication as often as before it.
    """

    pin_home_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)
    pin_trending_minutes: int | None = Field(default=None, ge=1, le=60 * 24 * 7)


class CmsArticleList(BaseModel):
    articles: list[CmsArticleOut]
    total: int


class DashboardStats(BaseModel):
    """§24 asks for 14 cards. The first seven are the editorial queue; the rest
    are audience and AI, which the dashboard previously did not show at all."""

    total_articles: int
    drafts: int
    pending_review: int
    approved: int
    published_today: int
    breaking_live: int
    returned_for_changes: int
    # §24 additions
    total_users: int = 0
    active_users_7d: int = 0
    views_today: int = 0
    submissions_pending: int = 0
    ai_suggestions_new: int = 0
    ai_drafts_pending: int = 0
    scheduled: int = 0
    top_categories: list[dict[str, Any]] = []
    top_districts: list[dict[str, Any]] = []
    top_mandals: list[dict[str, Any]] = []


class CmsOption(BaseModel):
    id: int
    slug: str
    name_te: str
    name_en: str


class CmsCategoryOption(CmsOption):
    parent_id: int | None = None


class CmsDistrictOption(CmsOption):
    state: str


class CmsEditorOptions(BaseModel):
    categories: list[CmsCategoryOption]
    districts: list[CmsDistrictOption]
    states: list[CmsOption] = []
    authors: list[dict[str, Any]] = []
    tags: list[CmsTagRef] = []

File: frontend/src/types/cms.ts
export interface CmsMediaRef { id:number; url:string; alt_te:string|null; credit:string|null; width:number|null; height:number|null }
export interface CmsVideoRef { id:number; youtube_id:string; title_te:string; thumbnail_url:string }
export interface CmsTagRef { id:number; slug:string; name_te:string; name_en:string }
/** §19 — the rendition the article uses. `provider:'upload'` is an editor's own file. */
export interface CmsAudioRef { id:number; url:string|null; mime:string; duration_sec:number; provider:string; status:string }
/** §8/§9 — a pin that is live right now, with its own countdown. */
export interface CmsActivePin { placement:'home'|'category'|'local'|'breaking'|'trending'; ends_at:string; seconds_remaining:number }

/** §23 — where the article came from, distinct from `source_type`'s copyright origin. */
export type ArticleType = 'NORMAL'|'REPORTER'|'USER_SUBMITTED'|'AI_SUGGESTED'|'AI_DRAFT'|'BREAKING_NEWS';

export interface CmsArticle {
  id:number; short_id:string; slug:string;
  title_te:string; title_en:string|null; sub_title_te:string|null; summary_te:string|null;
  body:Record<string,unknown>|null;
  category_id:number|null; subcategory_id:number|null;
  district_id:number|null; mandal_id:number|null; locality_id:number|null;
  author_id:number|null; byline_te:string|null;
  source_type:string; source_credit:string|null; article_type:ArticleType;
  status:string; workflow_state:string;
  is_breaking:boolean; is_exclusive:boolean; is_featured:boolean; voice_enabled:boolean;
  ai_generated:boolean;
  hero_media_id:number|null; video_id:number|null; audio_asset_id:number|null;
  seo_title:string|null; seo_description:string|null; canonical_url:string|null;
  approved_by:number|null; approved_at:string|null;
  published_at:string|null; scheduled_at:string|null; expires_at:string|null;
  breaking_until:string|null; updated_at:string;
  // §8/§9 placement chosen on the form; applied when the story goes live.
  pin_home_minutes:number|null; pin_trending_minutes:number|null;
  hero_media:CmsMediaRef|null; gallery:CmsMediaRef[]; video:CmsVideoRef|null; tags:CmsTagRef[];
  audio:CmsAudioRef|null; active_pins:CmsActivePin[];
}
export interface CmsArticleList { articles:CmsArticle[]; total:number }

/** §24 — fourteen cards plus the three "top" breakdowns. */
export interface DashboardStats {
  total_articles:number; drafts:number; pending_review:number; approved:number;
  published_today:number; breaking_live:number; returned_for_changes:number;
  scheduled:number; total_users:number; active_users_7d:number; views_today:number;
  submissions_pending:number; ai_suggestions_new:number; ai_drafts_pending:number;
  top_categories:Array<{id:number;count:number;name_te:string;name_en:string}>;
  top_districts:Array<{id:number;count:number;name_te:string;name_en:string}>;
  top_mandals:Array<{id:number;count:number;name_te:string;name_en:string}>;
}

export interface CmsOption { id:number; slug:string; name_te:string; name_en:string }
export interface CmsCategoryOption extends CmsOption { parent_id:number|null }
export interface CmsDistrictOption extends CmsOption { state:string }
export interface CmsEditorOptions {
  categories:CmsCategoryOption[]; districts:CmsDistrictOption[]; states:CmsOption[];
  authors:Array<{id:number;name_te:string;name_en:string}>; tags:CmsTagRef[];
}

/** §18 / §20 / §35 — the editable half of the settings screen. */
export interface SettingSpec { key:string; kind:string; default:unknown; description:string }
export interface SettingsPayload {
  environment:Record<string,unknown>;
  values:Record<string,unknown>;
  specs:SettingSpec[];
  voice_usage:{chars_this_month:number;monthly_budget:number;percent_used:number;assets_ready:number;assets_failed:number};
}

/** §16 / §17 — an AI story idea with its attribution. */
export interface AiSource { publisher:string; title:string|null; url:string; licence:string; excerpt:string|null }
export interface AiSuggestion {
  id:number; topic_te:string; topic_en:string|null; rationale_te:string|null;
  category_id:number|null; district_id:number|null; status:'new'|'accepted'|'rejected'|'used';
  score:number; engine:string; model:string|null; created_at:string;
  review_note:string|null; sources:AiSource[];
}
export interface AiDraft {
  id:number; suggestion_id:number|null; title_te:string; summary_te:string|null;
  body:Record<string,unknown>|null; body_plain:string|null;
  category_id:number|null; district_id:number|null;
  status:'draft'|'converted'|'discarded'; engine:string; model:string|null;
  confidence:number|null; word_count:number; article_id:number|null; created_at:string;
}

/** §19 — what the article page should do about voice. */
export interface AudioState {
  available:boolean; url:string|null; mime:string|null; duration_sec:number;
  voice:string|null; provider:string|null;
  /** 'device' means fall back to the browser voice; null means hide the player. */
  fallback:'device'|null; voice_enabled:boolean;
}

/** §17 — what right we hold over a source's words. */
export type SourceLicence =
  | 'agency_contract' | 'publisher_partner' | 'press_release'
  | 'government' | 'creative_commons' | 'own_network' | 'rss_public';
export type ContentPolicy = 'link_only' | 'excerpt_only' | 'full_text';
export type IngestStatus = 'new' | 'imported' | 'rejected' | 'duplicate';

export interface ContentSource {
  id: number; slug: string; name: string; name_te: string | null;
  feed_url: string; feed_kind: string;
  homepage_url: string | null; logo_url: string | null;
  licence: SourceLicence; content_policy: ContentPolicy;
  licence_note: string | null;
  /** The server's own verdict — never recompute it in the UI. */
  may_store_full_text: boolean;
  attribution_required: boolean; auto_publish: boolean;
  default_category_id: number | null; default_district_id: number | null;
  language: string; fetch_interval_minutes: number; is_active: boolean;
  last_fetched_at: string | null; last_status: string | null;
  consecutive_failures: number; items_ingested: number; pending_items: number;
  /** Hourly crawl configuration. */
  beat: SourceBeat;
  default_mandal_id: number | null;
  max_items_per_hour: number;
  allow_html_fallback: boolean;
  rewrite_enabled: boolean;
  mandal_autotag: boolean;
}

export type SourceBeat =
  | 'general' | 'national' | 'state' | 'district_local'
  | 'breaking' | 'sports' | 'film' | 'govt_jobs';

export type RewriteStatus =
  | 'none' | 'pending' | 'ready' | 'refused' | 'human_only' | 'skipped' | 'failed';

export type MandalMatchMethod =
  | 'none' | 'source_default' | 'keyword' | 'ambiguous' | 'editor';

export interface IngestedRewrite {
  id: number;
  status: RewriteStatus;
  title_te: string | null;
  summary_te: string | null;
  body_plain: string | null;
  attribution_te: string | null;
  word_count: number;
  engine: string;
  model: string | null;
  confidence: number;
  unverified: boolean;
  similarity_percent: number;
  refusal_reason: string | null;
  created_at: string;
}

/** Never an assignment — a guess, with its working shown. */
export interface MandalGuess {
  mandal_id: number | null;
  district_id: number | null;
  method: MandalMatchMethod;
  confidence: number;
}

export interface CrawlBeatStatus {
  beat: SourceBeat; quota: number; used: number; sources: number;
}

export interface CrawlStatus {
  enabled: boolean;
  rewrite_enabled: boolean;
  hourly_cap: number;
  used_this_hour: number;
  beats: CrawlBeatStatus[];
  last_fetch_at: string | null;
  /** No successful fetch for over two hours — usually a missing worker-ingest. */
  stale: boolean;
  queue: IngestQueueCounts;
}

export interface IngestedItem {
  id: number; title: string; summary: string | null;
  url: string | null; canonical_url: string | null;
  image_url: string | null; author: string | null;
  published_at: string | null; fetched_at: string;
  word_count: number; status: IngestStatus;
  article_id: number | null; review_note: string | null;
  source: { id: number; slug: string; name: string; licence: SourceLicence;
            content_policy: ContentPolicy; full_text: boolean } | null;
  has_full_text: boolean;
  mandal: MandalGuess;
  requires_human: boolean;
  rewrite_status: RewriteStatus;
  rewrite: IngestedRewrite | null;
}

export type IngestQueueCounts = Record<IngestStatus, number>;

export type AudioStatus = 'pending' | 'generating' | 'ready' | 'failed';

export interface AudioAssetRow {
  id: number;
  article_id: number;
  short_id: string | null;
  title_te: string | null;
  status: AudioStatus;
  provider: string;
  voice: string | null;
  language: string;
  duration_sec: number;
  char_count: number;
  bytes: number;
  /** Above 1 means long copy was synthesised in pieces and joined. */
  segment_count: number;
  error: string | null;
  generated_at: string | null;
  url: string | null;
}

export type BulletinStatus =
  | 'pending' | 'scripted' | 'ready' | 'published' | 'failed' | 'skipped';

export interface BulletinItemRef {
  position: number;
  article_id: number;
  short_id: string | null;
  /** The canonical reader path, built server-side. */
  url: string | null;
  headline_te: string;
}

export interface BulletinRow {
  id: number;
  date: string;
  slot: number;
  slot_label_te: string;
  status: BulletinStatus;
  revision: number;
  attempts: number;
  script_te: string | null;
  char_count: number;
  target_chars: number;
  duration_sec: number;
  segment_count: number;
  provider: string;
  voice: string | null;
  url: string | null;
  error: string | null;
  generated_at: string | null;
  published_at: string | null;
  items: BulletinItemRef[];
}

export interface BulletinList {
  date: string;
  enabled: boolean;
  requires_approval: boolean;
  items: BulletinRow[];
  /** Slots not yet produced, so the desk can offer "Run now". */
  missing_slots: number[];
}

export type KycStatus =
  | 'not_started' | 'draft' | 'submitted' | 'in_review'
  | 'more_info' | 'approved' | 'rejected' | 'expired';

export type ContributorType = 'citizen' | 'freelance' | 'student';

/** Metadata only. There is no url here, and there is none on the server. */
export interface KycDocumentRow {
  id: number;
  kind: string;
  mime: string;
  bytes: number;
  number_masked: string | null;
  uploaded_at: string | null;
  raw_path: string;
}

export interface KycProfileRow {
  id: number;
  user_id: number;
  name_te: string | null;
  phone: string | null;
  phone_verified: boolean;
  contributor_type: ContributorType | null;
  status: KycStatus;
  display_name_te: string;
  organisation: string | null;
  portfolio_url: string | null;
  course_year: number | null;
  submitted_at: string | null;
  reviewed_at: string | null;
  review_note: string | null;
  verified_badge: boolean;
  expires_at: string | null;
  provider: string;
  document_count: number;
  bio_te?: string | null;
  internal_note?: string | null;
  documents?: KycDocumentRow[];
  missing?: string[][];
}
